File: 129.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        return helper(root, 0);
    }
    
    int helper(TreeNode *root, int sum){
        if(!root){
            return 0;
        }
        sum = sum*10 + root->val;
        if(!root->left && !root->right){
            return sum;
        }
        return helper(root->left, sum) + helper(root->right, sum);
    }
};

File: golang/129/129.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sumNumbers(root *TreeNode) int {
	sum := 0
	helper(root, 0, &sum)
	return sum
}

func helper(root *TreeNode, currentSum int, sum *int) {
	if root == nil {
		return
	}

	currentSum = currentSum*10 + root.Val
	if root.Left == nil && root.Right == nil {
		*sum += currentSum
		return
	}

	helper(root.Left, currentSum, sum)
	helper(root.Right, currentSum, sum)
}

